src/spikeinterface/core/waveform_extractor.py

```diff
@@ -599,6 +577,40 @@ def _reset(self):
             # remove shared objects
             self._memory_objects = None
 
+    def set_recording(self, recording: Optional[BaseRecording], rec_attributes: Optional[dict] = None, allow_unfiltered: bool = False) -> None:
+        """
+
+        """
+
+        if recording is None:  # Recordless mode.
+            if rec_attributes is None:
+                raise ValueError("WaveformExtractor: if recording is None, then rec_attributes must be provided.")
+            for k in ('channel_ids', 'sampling_frequency', 'num_channels'):  # Some check on minimal attributes (probegroup is not mandatory)
+                if k not in rec_attributes:
+                    raise ValueError(f"WaveformExtractor: Missing key '{k}' in rec_attributes")
+            for k in ('num_samples', 'properties', 'is_filtered'):
+                if k not in rec_attributes:
+                    warn(f'Missing optional key in rec_attributes {k}: '
+                         f'some recordingless functions might not be available')
+        else:
+            if recording.get_num_segments() != self.get_num_segments():
+                raise ValueError(f"Couldn't set the WaveformExtractor recording: num_segments do not match!\n{self.get_num_segments()} != {recording.get_num_segments()}")
+            if abs(recording.sampling_frequency - self.sampling_frequency) > 1e-2:
+                raise ValueError(f"Couldn't set the WaveformExtractor recording: sampling frequency doesn't match!\n{self.sampling_frequency} != {recording.sampling_frequency}")
+            try:  # Will fail if called from __init__ because can't check from previous channel_ids.
+                if np.array_equal(self.channel_ids, recording.channel_ids):
+                    raise ValueError(f"Couldn't set the WaveformExtractor recording: channel_ids do not match!\n{self.channel_ids} != {recording.channel_ids}")
+            except:
+                pass
```

Comment:
Maybe check this only `if rec_attributes is not None`?

Reply:
You mean you could set a new recording with new attributes (like new `channel_ids`) that are different from the previous recording?

Wouldn't that break things?

Reply:
The `__init__` is mainly called by the [`cls.create()`](https://github.com/SpikeInterface/spikeinterface/blob/main/src/spikeinterface/core/waveform_extractor.py#L302) and the [`load()`](https://github.com/SpikeInterface/spikeinterface/blob/main/src/spikeinterface/core/waveform_extractor.py#L184) functions. In both cases, the `rec_attributes` are given. As you mentioned, the `self.channel_ids` cannot be accessed if the object is not initialized aready, so it shouldn't be used in the function.

If you call the `set_recording()` after the object has been created with `with_recording=False`, then the `self._rec_attributes` is already set.

So here's what I would do:

In the `__init__`:

```
...
self._rec_attributes = None
self.set_recording(recording, rec_attributes, allow_unfiltered) # this will set `self._rec_attributes`
```

in the `set_recording()`:
```
if self._rec_attributes is not None:
    channel_ids = self._rec_attributes['channel_ids']
elif rec_attributes is not None:
    channel_ids = rec_attributes['channel_ids']
else:
    raise ValueError
if not np.array_equal(channel_ids, recording.channel_ids):
    raise ValueError(...)
```